### PR TITLE
fix bug #1215

### PR DESCRIPTION
**Pull request by gaozilai**
fix bug #1215

### CODE REVIEW

File: src/lib_json/json_reader.cpp [hunk 1]
```diff
@@ -1417,6 +1417,8 @@ bool OurReader::readString() {
   Char c = 0;
   while (current_ != end_) {
     c = getNextChar();
+    if (c < 0)
```
Comment (dota17) [inline src/lib_json/json_reader.cpp:1420]: This check isn't make any sense, type `char` will not less than zero.

Comment (gaozilai) [reply src/lib_json/json_reader.cpp:1420]: Are you sure? What about `unsigned char`?

Actually, there was an error report without that. And it works fine with it.

Comment (BillyDonahue) [reply src/lib_json/json_reader.cpp:1420]: > This check isn't make any sense, type `char` will not less than zero.


the `Reader::Char` typedef is just char. It is implementation defined whether char is signed or unsigned. I agree it's concerning to see it compared with c<0. On some platforms it's impossible. Others it's fine. So it's not portable code.

Note that char, signed char, and unsigned char are three distinct types in the C++ type system.

(chiming in from myphone)

Comment (gaozilai) [reply src/lib_json/json_reader.cpp:1420]: In GBK or GB2312, Chinese characters are composed of two BYTEs, A BYTE greater than 127 followed by any BYTE.

BYTE is unsigned char.
`BYTE` than 127 is a `char` less than 0.

Comment (BillyDonahue) [reply src/lib_json/json_reader.cpp:1420]: > In GBK or GB2312, Chinese characters are composed of two BYTEs, A BYTE greater than 127 followed by any BYTE.
> 
> 
> 
> BYTE is unsigned char.
> 
> `BYTE` than 127 is a `char` less than 0.
> 
> 

Ok there's your problem. We don't decode GBK. That's not a valid JSON string encoding. You will have to convert to UTF-8 before feeding to a JsonCpp Reader.

BYTE may have gone out as an unsigned char, but isn't unsigned char unless the receiver doing the decoding decides to look at it as an unsigned char.

Comment (gaozilai) [reply src/lib_json/json_reader.cpp:1420]: Yes, you are very correct! Thank you!

**Comment by dota17** [comment]
Thank you for your patience:) @gaozilai 
By the way, you should avoid making any changes in `master` branch, keep it clean and keep it track with upstream master branch, then if you want to submit PR, please checkout a new `dev` branch from the master branch. 
And if you've made some changes of code, it's better to write some test cases to make sure your changes really work.

**Comment by gaozilai** [comment]
> Thank you for your patience:) @gaozilai
> By the way, you should avoid making any changes in `master` branch, keep it clean and keep it track with upstream master branch, then if you want to submit PR, please checkout a new `dev` branch from the master branch.
> And if you've made some changes of code, it's better to write some test cases to make sure your changes really work.

ok, got it. (Although I don't know how to do yet）
I am a svn user, work on the stand-alone machine. I learned so much, I knew nothing about git and json a week ago.
My practice is over, I will not make any changes anymore.

`jsoncpp` is perfect!
thank you

**Pull request closed without merging**